### PR TITLE
Allow plugins to install bootstrap checks

### PR DESCRIPTION
**Pull request by s1monw**
Plugins also have the need to provide better OOTB experience by configuring
defaults unless the plugin is used in _production_ mode. This change exposes
the bootstrap check infrastructure as part of the plugin API to allow plugins
to specify / install their own bootstrap checks if necessary.

### CODE REVIEW

**Comment by dadoonet** [comment]
Lovely. 

Do you think it is worth adding some doc about `getBootstrapChecks()` in our [developer guide](https://www.elastic.co/guide/en/elasticsearch/plugins/current/plugin-authors.html)?

**Comment by s1monw** [comment]
> Do you think it is worth adding some doc about getBootstrapChecks() in our developer guide?

I am not sure, there are javadocs for this on the `Plugin.java` class. I wonder if we should really duplicate efforts? I think we can add a reference to [this](https://www.elastic.co/guide/en/elasticsearch/reference/master/bootstrap-checks.html) that plugins can also add those checks?

**Comment by dadoonet** [comment]
>  I think we can add a reference to this that plugins can also add those checks?

That would be nice indeed so folks who are reading docs won't be surprised if a plugin barks at startup.

**Review by dadoonet — APPROVED** [review]
Very nice.

As you said, can you add some doc in bootstrap checks?

**Comment by s1monw** [comment]
@elasticmachine test this please

File: core/src/main/java/org/elasticsearch/bootstrap/Check.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.bootstrap;
+
+/**
+ * Encapsulates a bootstrap check.
+ */
+public interface Check {
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/bootstrap/Check.java]: I wonder if this class should be renamed to `BootstrapCheck` and the currently-existing `BootstrapCheck` should be renamed to `BootstrapChecks` (or something like that, basically looking to give `Check` a more indicative name now that it is not an nested class in `BootstrapCheck`)?

**Review by jasontedor — APPROVED** [review]
I left one suggestion, but it LGTM either way.

**Comment by s1monw** [comment]
@elasticmachine test this please

**Pull request merged**